subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/CrossVersionTestEngine.java

```diff
@@ -0,0 +1,234 @@
+        if (crossVersionTest == null) {
+            return new EngineDescriptor(uniqueId, "skip");
+        }
+
+        buildTestVariants(uniqueId, discoveryRequest);
+        return discoverTests(uniqueId);
+    }
+
+    @Override
+    protected SpockExecutionContext createExecutionContext(ExecutionRequest request) {
+        return new SpockExecutionContext(request.getEngineExecutionListener());
+    }
+
+    private void buildTestVariants(UniqueId rootId, EngineDiscoveryRequest discoveryRequest) {
+        variants.add(TestVariant.tapiCurrent(rootId, discoveryRequest));
+        variants.add(TestVariant.selected(rootId, discoveryRequest));
+        variants.add(TestVariant.tapiTargetLoaded(rootId, discoveryRequest));
+    }
+
+    private TestDescriptor discoverTests(UniqueId uniqueId) {
+        EngineDescriptor rootDescriptor = new SpockEngineDescriptor(uniqueId, RunContext.get());
+        for (TestVariant testVariant : variants) {
+            for (TestDescriptor test : testVariant.discover(delegateEngine).getChildren()) {
+                rootDescriptor.addChild(test);
+            }
+        }
+        return rootDescriptor;
+    }
+}
+
+class TestVariant {
+    private final UniqueId id;
+    private final EngineDiscoveryRequest discoveryRequest;
+    private final Map<String, String> systemProperties = new HashMap<String, String>();
+
+    static TestVariant tapiCurrent(UniqueId rootId, EngineDiscoveryRequest discoveryRequest) {
+        TestVariant testVariant = new TestVariant(rootId, "tapiCurrent", new ToolingApiCurrentDiscoveryRequest(discoveryRequest));
+        testVariant.systemProperties.put(VERSIONS_SYSPROP_NAME, CoverageContext.DEFAULT.selector);
+        return testVariant;
+    }
+
+    static TestVariant selected(UniqueId rootId, EngineDiscoveryRequest discoveryRequest) {
+        return new TestVariant(rootId, "selected", discoveryRequest);
+    }
+
+    static TestVariant tapiTargetLoaded(UniqueId rootId, EngineDiscoveryRequest discoveryRequest) {
+        TestVariant testVariant = new TestVariant(rootId, "tapi", new ToolingApiClassloaderDiscoveryRequest(discoveryRequest));
+        testVariant.systemProperties.put("org.gradle.integtest.currentVersion", GradleVersion.current().getVersion());
+        return testVariant;
+    }
+
+    private TestVariant(UniqueId rootId, String variant, EngineDiscoveryRequest request) {
+        this.id = rootId.append("variant", variant);
+        this.discoveryRequest = request;
+    }
+
+    TestDescriptor discover(final TestEngine discoveryEngine) {
+        return SystemProperties.getInstance().withSystemProperties(systemProperties, new Factory<TestDescriptor>() {
+            @Override
+            public TestDescriptor create() {
+                return discoveryEngine.discover(discoveryRequest, id);
+            }
+        });
+    }
+}
+
+class DelegatingDiscoveryRequest implements EngineDiscoveryRequest {
```

Comment:
This interface is not really supposed to be implemented by clients. I think it will work for now but we should open an issue for JUnit to provide an API to create a new request based on an existing one. If we go for `ClassSelectors` with `Strings`, we wouldn't need it either way.

Reply:
> If we go for ClassSelectors with Strings, we wouldn't need it either way.

I am not sure this would be enough - the `ClassSelector` will load the class on the first discovery iteration using whatever classloader is set in the current thread. And here I want to run the discovery multiple times, possibly using a different classloader for a particular iteration, which would not work if the classes are already loaded in that discovery request during a previous iteration. 

Perhaps then instead of implementing a custom `EngineDiscoveryRequest`, I can create new ones from existing ones using newly created `ClassSelectors`.

Reply:
You're right, I forgot that `ClassSelector` caches the loaded class. That maybe wasn't such a wise decision in hindsight. Carry on then... 🙂